hello-world-with-widgets/index.ts

```diff
@@ -0,0 +1,287 @@
+  50000: '50k - 100k',
+  100000: '100k - 150k',
+  150000: '150k - 200k',
+  200000: '200k - 250k',
+  250000: '250k - 300k',
+  300000: '300k - 350k',
+  350000: '350k - 400k',
+  400000: '400k - 450k',
+  450000: '450k - 500k',
+  500000: '> 500k'
+};
+
+// SPATIAL FILTER
+// prepare a function to get the new viewport state, that we'll pass debounced to our widgets to minimize requests
+
+const debouncedRenderWidgets = debounce(renderWidgets, 500);
+
+// sync deckgl map after user interaction, obtain new viewport after
+
+let viewportSpatialFilter;
+
+deck.setProps({
+  onViewStateChange: ({viewState}) => {
+    const {longitude, latitude, ...rest} = viewState;
+    map.jumpTo({center: [longitude, latitude], ...rest});
+    const viewport = new WebMercatorViewport(viewState);
+    viewportSpatialFilter = createViewportSpatialFilter(viewport.getBounds());
+    formulaWidget.innerHTML = 'Loading...';
+    histogramWidgetChart.showLoading();
+    debouncedRenderWidgets();
+  }
+});
+
+// COLUMN FILTERS
+// prepare filters
+
+let filters = {};
+
+// listen to inputs from HTML
+
+let selectedCountry = 'All';
+const countrySelector = document.querySelector<HTMLSelectElement>('#osmCategorySelector');
+
+countrySelector?.addEventListener('change', () => {
+  const country = countrySelector.value;
+  if (country == 'All') {
+    clearFilters(filters);
+  } else {
+    addFilter(filters, {
+      column: 'sov_a3',
+      type: FilterType.IN,
+      values: [country]
+    });
+  }
+  selectedCountry = countrySelector.value;
+  initialize();
+});
+
+// filtering from the widget, including a clear filters button
+
+const clearFiltersButton = document.querySelector<HTMLElement>('#clearHistogramFilter');
+
+clearFiltersButton?.addEventListener('click', () => {
+  clearHistogramFilter();
+});
+
+function filterViaHistogram(dataIndex) {
+  clearFiltersButton.style.display = 'inherit';
+
+  const histogramKeys = Object.keys(histogramLabels).map(Number);
+  const minValue = histogramKeys[dataIndex];
+  const maxValue = histogramKeys[dataIndex + 1];
+
+  removeFilter(filters, {
+    column: 'pop_max'
+  });
+
+  if (dataIndex === histogramKeys.length - 1) {
+    // For the last category (> 1M), use CLOSED_OPEN
+    addFilter(filters, {
+      column: 'pop_max',
+      type: FilterType.CLOSED_OPEN,
+      values: [[minValue, Infinity]]
+    });
+  } else {
+    // For first and middle categories, use BETWEEN
+    addFilter(filters, {
+      column: 'pop_max',
+      type: FilterType.BETWEEN,
+      values: [[minValue, maxValue]]
+    });
+  }
+
+  initialize();
+}
+
+function clearHistogramFilter() {
+  clearFiltersButton.style.display = 'none';
+
+  removeFilter(filters, {
+    column: 'pop_max'
+  });
+
+  initialize();
+}
+
+// RENDER
+// render Widgets function
+
+async function renderWidgets() {
+  // configure widgets
+
+  const formula = await dataSource.widgetSource.getFormula({
+    operation: 'count',
+    spatialFilter: viewportSpatialFilter
+  });
```

Comment:
Sometimes after reloading the page, I'm seeing an error thrown here because `dataSource` is not yet defined. I can reproduce the issue by trying to scroll/zoom with the mousewheel before tiles load, but maybe there are other possible ways it could happen.

Maybe the easiest fix is to check and exit early if `dataSource` is not ready yet? We could use [signals](https://github.com/preactjs/signals) or something like that, but it feels a bit framework-ey and probably these examples are trying to avoid that.

Reply:
Another option would be to make sure that `dataSource` is always a Promise (never undefined), and the functions that depend on it would just always do an `await` before using it.

Reply:
Dsicussed and fixed in https://github.com/CartoDB/deck.gl-examples/pull/27/commits/67133629a98006b39a9c885f9039c935d357d0dd — Thanks! :)